bftengine/src/bftengine/BftExecutionEngine.cpp

```diff
@@ -0,0 +1,406 @@
+        static_cast<uint32_t>(config_.getmaxReplyMessageSize() - sizeof(ClientReplyMsgHeader)),
+        (char*)std::malloc(config_.getmaxReplyMessageSize() - sizeof(ClientReplyMsgHeader)),
+        req.requestSeqNum(),
+        req.result()};
+    // Decode the pre-execution block-id for the conflict detection optimization,
+    // and pass it to the post-execution.
+    if (req.flags() & HAS_PRE_PROCESSED_FLAG) {
+      ConcordAssertGT(req.requestLength(), sizeof(uint64_t));
+      auto requestSize = req.requestLength() - sizeof(uint64_t);
+      auto opt_block_id = *(reinterpret_cast<uint64_t*>(req.requestBuf() + requestSize));
+      LOG_DEBUG(GL, "Conflict detection optimization block id is " << opt_block_id);
+      execution_request.blockId = opt_block_id;
+      execution_request.requestSize = requestSize;
+    }
+    accumulatedRequests.push_back(execution_request);
+  }
+  return accumulatedRequests;
+}
+void BftExecutionEngineBase::addPostExecCallBack(
+    std::function<void(PrePrepareMsg*, IRequestsHandler::ExecutionRequestsQueue&)> cb) {
+  post_exec_handlers_.add(std::move(cb));
+}
+void BftExecutionEngineBase::execute(std::deque<IRequestsHandler::ExecutionRequest>& accumulatedRequests,
+                                     Timestamp& timestamp) {
+  const std::string cid = accumulatedRequests.back().cid;
+  const auto sn = accumulatedRequests.front().executionSequenceNum;
+  concordUtils::SpanWrapper span_wrapper{};
+  LOG_INFO(getLogger(), "Executing all the requests of preprepare message: " << KVLOG(cid, sn));
+  concord::diagnostics::TimeRecorder scoped_timer1(*metrics_.executeWriteRequest);
+  requests_handler_->execute(accumulatedRequests, timestamp, cid, span_wrapper);
+  if (accumulatedRequests.size() == 1) timestamp.request_position++;
+}
+void BftExecutionEngineBase::loadTime(SeqNum sn) {
+  if (config_.timeServiceEnabled) {
+    timestamps[sn].time_since_epoch = time_service_manager_->compareAndUpdate(timestamps[sn].time_since_epoch);
+    LOG_INFO(getLogger(),
+             "Timestamp to be provided to the execution: " << timestamps[sn].time_since_epoch.count() << "ms");
+  }
+}
+
+using namespace concord::diagnostics;
+
+class BlockAccumulationExecutionEngine : public BftExecutionEngineBase {
+ public:
+  BlockAccumulationExecutionEngine(std::shared_ptr<IRequestsHandler> requests_handler,
+                                   std::shared_ptr<ClientsManager> client_manager,
+                                   std::shared_ptr<ReplicasInfo> reps_info,
+                                   std::shared_ptr<PersistentStorage> ps,
+                                   std::shared_ptr<TimeServiceManager<CLOCK_TYPE>> time_service_manager)
+      : BftExecutionEngineBase{requests_handler, client_manager, reps_info, ps, time_service_manager} {}
+  SeqNum addExecutions(const std::vector<PrePrepareMsg*>& ppMsgs) override {
+    SeqNum sn = 0;
+    for (auto ppMsg : ppMsgs) {
```

Comment:
Is there a need to make a copy here?
If not, change to const ref.

Reply:
The objects in the vector are pointers, so no need for references (the reason for using raw pointers, is that currently changing it to some smart pointer will take a lot of efforts, I would prefer to avoid this now)

Reply:
Then using auto* will make it clearer, I know that the type is in the signature, but it saves the reader the need to go there to understand whether making a copy makes sense.

Reply:
done